### PR TITLE
Fix asynchronous read/writes to seriesCache

### PR DESCRIPTION
**Pull request by JoelSpeed**
This map is read from/written to within separate go routines and so needs some sort of mutex. Have been testing this fix and it has stopped the periodic crashing that we have been seeing

### CODE REVIEW

File: pkg/custom-provider/provider.go [hunk 1]
```diff
@@ -298,15 +301,19 @@ func (l *cachingMetricsLister) updateMetrics() error {
 			errs <- nil
 			continue
 		}
+		seriesCacheByQueryLock.Lock()
```
Comment (DirectXMan12) [inline pkg/custom-provider/provider.go]: wouldn't you need a lock on the read above as well?

**Comment by JoelSpeed** [comment]
I think there are 3 ways to solve this:
- Use the locks before and after every read write as (attempted) in this PR
- Use a `sync.Map` to replace the existing map - effectively the same as above but has all of the locking wrapped into the `Load` and `Store` methods so looks neater
- Create a list of selectors in one loop, create a channel based on the length of the list of selectors, spawn go routines in another loop and send the result onto the channel. Then move into the map from the channel. With this approach the channel would need to take structs containing both the selector and the series to enable them to be mapped in the final loop, is this perhaps a little complex?

I think the cleanest is the second option with the `sync.Map`, although I haven't used them before so have no idea if that's actually a good idea

**Comment by JoelSpeed** [comment]
I pushed a second branch that now uses the `sync.Map` method. I've been running this on a testing cluster for the last few days and haven't seen any concurrent RW issues. Do you think this could be a good solution or would you rather try the channel based approach? https://github.com/pusher/k8s-prometheus-adapter/tree/feature/advanced-config2

**Comment by JoelSpeed** [comment]
I've added a commit to change this PR to send the required data down a channel to avoid the concurrency issue. I think this should be safe now?

**Comment by JoelSpeed** [comment]
#46 was merged so this is no longer needed

**Pull request closed without merging**